javascript/react/introduction-to-hooks/why-were-hooks-needed.md

```diff
@@ -0,0 +1,81 @@
+---
+author: kapnobatai136
+
+levels:
+
+  - medium
+  
+aspects:
+
+  - new
+
+  - workout
+
+  - deep
+
+type: normal
+
+category: must-know
+
+links:
+  - '[render-props](https://reactjs.org/docs/render-props.html){documentation}'
+  - '[higher-order-components](https://reactjs.org/docs/higher-order-components.html){documentation}'
+
+---
+
+# Why were hooks needed?
+
+---
+## Content
+
+Hooks represent a new addition to React, and they allow you to use state or other useful features without using class components.
+
+Hooks were created to solve a number of issues that were encountered in the past five years since React was created.
+
+In React, there is no direct way to "attach" reusable behavior to a component (such as connecting to a store). The patterns that emerged from the need of reusable behavior were either using [render props](https://reactjs.org/docs/render-props.html) or using [higher-order components](https://reactjs.org/docs/higher-order-components.html) (or HOCs).
+
+A `render prop` is nothing more than a function that a component uses to know what to render.
+
+```jsx
+<MyComponent
+  render={person => (
+    <p>Hello {person.name}</p>
+  )}
```

Comment:
Add a note here that another version of the "render props" pattern is to pass a function as `children`:

```jsx
<MyComponent>
  {person => (
    <p>Hello {person.name}</p>
  )}
</MyComponent>
```